Refactor sockets and hooks

haotongwang opened the pull request:
## Description

Sockets

- Incoming status MQTT payloads are stored in a global object with dynamic structure
- The "subcomponent" object is emitted every incoming payload
- Client can request part of the status payloads stored on the server using the channel `get-status-payload`

Hooks

- Lots of repeated code between hooks used to get retained payloads
- Use a general `createStatusPayloadHook` function to create the hooks

## Steps to Test

Check that the test cases for previous iterations also pass for this update.

haotongwang committed:
add util functions

haotongwang committed:
update packages

haotongwang committed:
update eslint

haotongwang committed:
sockets for server side

haotongwang committed:
update packages; fix reference bug

haotongwang committed:
use new socket stuff

haotongwang committed:
refactor hooks

hallgchris commented on server/js/sockets.js:
```diff
@@ -177,30 +169,48 @@ sockets.init = function socketInit(server) {
-            payloadString,
-          );
+        // topicString: ["", "v3", "status", "<component>", "<subcomponent>", ...properties]
+        const value = JSON.parse(payloadString);
```
This will cause a crash if we somehow receive invalid JSON. e.g.
```bash
mosquitto_pub -t "/v3/status" -m ""
```
Would be good to handle this more neatly, even if it's just printing an error and dropping the message.

haotongwang replied:
Done
411e617

Message gets dropped and an error is printed to console

hallgchris commented on server/js/sockets.js:
```diff
@@ -177,30 +169,48 @@ sockets.init = function socketInit(server) {
-        // Send mqtt payload on corresponding channel
+        // Emit subcomponent
+        const component = topicString[3];
+        const subcomponent = topicString[4];
```
Doesn't cause a crash, but also may as well handle when these two strings are undefined. (currently it emits on `status-undefined-undefined` which isn't harmful but should be an easy fix)

haotongwang replied:
Done
411e617

Replaces `undefined` with an empty string so topics like `status--subcomponent` are valid

hallgchris commented on server/js/sockets.js:
```diff
@@ -177,30 +169,48 @@ sockets.init = function socketInit(server) {
         socket.emit(
-          `status-${component}-${subComponent}`,
-          global.statusPayloads[component][subComponent],
+          `status-${component}-${subcomponent}`,
+          retained.status?.[component]?.[subcomponent],
+        );
+        console.log(
+          `received emit status-${component}-${subcomponent}: ${JSON.stringify(
+            retained.status?.[component]?.[subcomponent],
+          )}\n`,
```
Would be neat to extract the repeated code here to variables

haotongwang replied:
The `console.log` will be removed before this get's merged. I'm only leaving it there so I can debug stuff easier

khanguslee replied:
@haotongwang Perhaps the next task you can work on is to add an actual logger to dashboard? The logger would be enabled for DEBUG but disabled in prod via env variables. Thoughts on this @hallgchris ? I've been meaning to add this..

hallgchris commented on server/js/util.js:
```diff
@@ -0,0 +1,82 @@
+    return getPropWithPath(nextObj, path.slice(1));
+  };
+
+  const setPropWithPath = (obj, path, value) => {
```
Some docs on this method would be good!

haotongwang replied:
Done
411e617

hallgchris commented on server/js/util.js:
```diff
@@ -0,0 +1,82 @@
+module.exports = (function util() {
```
For consistency, could you use the same style module exports that we use elsewhere?

Side question, is this your original code? If not, would be good to link the source.

haotongwang replied:
Done
cbc5c49

Yes it is my original code. I only put it in a module because it was cluttering up the sockets file.

hallgchris commented on server/js/util.js:
```diff
@@ -0,0 +1,82 @@
+  //     );
+
+  //   return deepMerge({ ...target, [key]: val })(noProp(source, key));
+  // };
```
I'd rather avoid having large chunks of commented code like this - if we're going to need it in the future leave it uncommented, otherwise we shouldn't need it. (Also, a multiline comment would be neater if we were to keep it)

haotongwang replied:
Done
411e617

hallgchris commented on server/js/sockets.js:
```diff
@@ -177,30 +169,48 @@ sockets.init = function socketInit(server) {
-          );
+        // topicString: ["", "v3", "status", "<component>", "<subcomponent>", ...properties]
+        const value = JSON.parse(payloadString);
+        const props = topicString.slice(2);
```
I find the variable names `value` and `props` a little confusing... I reckon `payload` would be more clear for the first, and maybe could you add a comment describing what `props` is? (maybe `path` would be a better name, as it's usage is similar to what you've called `path` elsewhere below?

haotongwang replied:
`payload` is already defined previously so I used `value` which is consistent with the usage in `propsToObj` and `setPropWithPath` as the payload is the value at the end of the path.

Changed `props` to `path`
0497ffb

hallgchris commented on server/.eslintrc:
```diff
@@ -1,7 +1,17 @@
+    "one-var": "error",
+    "no-plusplus": "off",
+    "no-await-in-loop": "off",
+    "no-console": "off",
```
Was `eslint` complaining about our code?

haotongwang replied:
Half of them were from complaining about `server.js`, I added a couple like `one-var` for consistency.

hallgchris commented on server/js/sockets.js:
```diff
@@ -177,30 +169,48 @@ sockets.init = function socketInit(server) {
     });
 
+    socket.on('get-status-payload', (path) => {
+      console.log(retained.status);
+      console.log(path);
+      console.log(
+        `status-${path.join('-')}: ${JSON.stringify(
+          getPropWithPath(retained.status, path),
+        )}\n`,
+      );
```
These comments were really great for seeing how your PR worked, but they seem very verbose to have running all the time. IMO we should probably remove them, unless you had a reason otherwise

haotongwang replied:
The console.log will be removed before this get's merged. I'm only leaving it there so I can debug stuff easier.

I should have specified that with a comment

hallgchris commented on server/js/sockets.js:
```diff
@@ -177,30 +169,48 @@ sockets.init = function socketInit(server) {
+      );
+      socket.emit(
+        `status-${path.join('-')}`,
+        getPropWithPath(retained.status, path),
```
Can we only emit if `path` was valid and returned data?

haotongwang replied:
Done
33a28c0

I really need to remember to validate my inputs...

hallgchris commented on server/js/sockets.js:
```diff
@@ -177,30 +169,48 @@ sockets.init = function socketInit(server) {
+      );
+    });
+
     // TODO: Fix up below socket.io handlers
```
@khanguslee what about these needs "fixing up"?

khanguslee replied:
What a throwback. Not sure what needed fixing up tbh. However, looking at those event names they aren't too great and should probably follow an API-like style. Not really necessary tho.

haotongwang replied:
I think we'll be refactoring most of this stuff for V3 anyway. It works atm so I guess it's fine?

hallgchris commented on client/src/components/v2/CameraRecordingStatus.js:
```diff
@@ -19,6 +19,8 @@ const defaultStatus = 'Waiting for status...';
 export default function CameraRecordingStatus({ device }) {
   const payload = useCameraRecordingStatus(device);
 
+  console.log(`recording status: ${payload}`);
```
Again, good for debugging this PR but probably don't need this in production.

haotongwang replied:
The console.log will be removed before this get's merged. I'm only leaving it there so I can debug stuff easier.

I should have specified that with a comment

hallgchris commented on client/src/api/v2/camera.js:
```diff
@@ -126,33 +117,10 @@ function parseRecordingPayloadData(data) {
-
-  return parseRecordingPayloadData(lastPayload && lastPayload[device]);
+function initStatus(subComponent, device) {
+  emit(`get-status-payload`, ["camera", subComponent].concat(device ? [device] : []));
```
Putting stuff on one line sure is satisfying but I think this could be more clear if the logic was split up a bit more.

haotongwang replied:
Done
8e26e76

My poor one liners :(

hallgchris commented on client/src/api/v2/camera.js:
```diff
@@ -204,19 +211,9 @@ export function useVideoFeedStatus() {
+export const useCameraStatus = (device) => createStatusPayloadHook(
+  device, {
+    initValue: false,
+    payloadHandler: (setState) => (newPayload) => setState(newPayload?.connected ?? false)
```
This is kinda hurting my brain... Can you explain what's going on here?

khanguslee replied:
Its the very latest javascript operators. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

khanguslee replied:
Side note. @haotongwang definitely does functional programming LOL

khanguslee replied:
https://medium.com/javascript-scene/curry-and-function-composition-2c208d774983

khanguslee replied:
I definitely feel like curry functions are an advanced technique (you're just too good @haotongwang). The only place that I HAVE seen curried functions is when you work with selectors via react + redux. In other normal usage, I haven't seen it used, purely because you will get the same reaction as @hallgchris did. Really up to the team to decide if this is something we want to use within the codebase.

haotongwang replied:
@khanguslee All from FIT2102 haha. I can change it to not use curried functions but it just makes the code less clean but more understandable for people not familiar with curried functions. I'll see if I can remove the curried functions without making the implementation too messy.

khanguslee replied:
Curious as to what the implementation looks like without curried functions @haotongwang

haotongwang replied:
I think the new operators are useful and makes it much more understandable (e.g. `p ?? initValue` would have to be replaced with `p === null && initvalue`.

I've removed the curried function for `payloadHandler`. The code's a bit messier but I agree curried functions might be a bit too much...
6e427c4

hallgchris commented on client/src/api/v2/camera.js:
```diff
@@ -126,33 +117,10 @@ function parseRecordingPayloadData(data) {
-
-  return parseRecordingPayloadData(lastPayload && lastPayload[device]);
+function initStatus(subComponent, device) {
+  emit(`get-status-payload`, ["camera", subComponent].concat(device ? [device] : []));
```
Since this is going to be used for more than just camera stuff in the future - should it be made more generic and moved to another/a new file?

haotongwang replied:
I think it's fine to leave it here for now. I think of it like `initStatus` for `camera` (thus hardcoding the "camera" part as the component). The `status.js` API currently is for the sensors so it would be a bit awkward to fit it in. Refactoring it to be more generic should be quite easy anyway.

hallgchris reviewed (changes requested):
There's some super cool code here, and I'll be honest, some of it I haven't quite got my head around and may need you to explain :joy: would be good to know what someone more experienced thinks about it - to me it's quite complex and might be a bit too much so for our "beginner friendly" web dev project, but maybe I'm just too beginner.

That aside though, this is a much much cooler neater way to handle socket stuff!

hallgchris commented:
Just realised that the video feed status isn't working. Does it work for you?
![image](https://user-images.githubusercontent.com/17876556/94010200-f3a2e380-fde8-11ea-8b5c-83b19f9961ec.png)

haotongwang commented:
> Just realised that the video feed status isn't working. Does it work for you?
> ![image](https://user-images.githubusercontent.com/17876556/94010200-f3a2e380-fde8-11ea-8b5c-83b19f9961ec.png)

That looks like it's working for me? It's meant to be on when you can start recording. If you're talking about the camera status, then that turns on when `orchestrator.py` is running.

haotongwang committed:
Clean up code

Add exception handling for parsing payload
Add undefined handling for variables
Add function headers
Remove unused commented function

haotongwang committed:
Unwrap IIFE in utils

haotongwang committed:
Rename props to path

haotongwang committed:
Add input validation for get-status-payload handler

haotongwang committed:
move initStatus logic to multiple lines

haotongwang committed:
Remove curried function :(

haotongwang committed:
Fix linting

hallgchris reviewed (approved):
Looks good! No clue what was going through my head before. I was either running the wrong thing or looking at the wrong place. Seems to be working fine.
